### PR TITLE
sync_diff_inspector: make exclude table support regex

### PR DESCRIPTION
**Pull request by lichunzhu**
### What problem does this PR solve? 
When we try to exclude some tables, we need to write them one by one which is inconvenient.

### What is changed and how it works?
make `check-tables.exclude_table` support regex

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)
```toml
[[check-tables]]
schema = "chauncy"
tables = ["~.*"]
exclude-tables = ["~t[2-3]"]
```
Run with this config and t2, t3 are not equal. diff can pass.

Related changes

 - Need to update the documentation

### CODE REVIEW

File: sync_diff_inspector/diff.go [hunk 1]
```diff
@@ -326,12 +326,11 @@ func (df *Diff) AdjustTableConfig(cfg *Config) (err error) {
...
-				if df.InExcludeTables(schemaTables.ExcludeTables, t) {
-					continue
-				} else {
-					tables = append(tables, t)
-				}
+			for _, excludeTable := range schemaTables.ExcludeTables {
+				matchedTables = df.RemoveMatchTable(excludeTable, matchedTables)
+			}
+			for table := range matchedTables {
+				tables = append(tables, table)
```
Comment (glorv) [inline sync_diff_inspector/diff.go:336]:
```suggestion
			tables = append(tables, matchedTables...)
```

File: sync_diff_inspector/diff.go [hunk 2]
```diff
@@ -537,6 +536,25 @@ func (df *Diff) GetMatchTable(db DBConfig, schema, table string, allTables map[s
 	return tableNames, nil
 }
 
+// RemoveMatchTable removes all the matched table.
+func (df *Diff) RemoveMatchTable(table string, allTables map[string]interface{}) map[string]interface{} {
```
Comment (glorv) [inline sync_diff_inspector/diff.go]: I think we don't need to depend on an `map[string]interface{}` only to do this remove operation, just use slice can do the same thing.

File: sync_diff_inspector/diff.go (hunk 1 above)
Comment (glorv) [reply sync_diff_inspector/diff.go:336]: If we don't change matchTables from  `slice` to `map`

File: sync_diff_inspector/diff.go [hunk 3]
```diff
@@ -537,6 +536,25 @@ func (df *Diff) GetMatchTable(db DBConfig, schema, table string, allTables map[s
...
+	tableNames := allTables
+
+	if table[0] == '~' {
+		tableRegex := regexp.MustCompile(fmt.Sprintf("(?i)%s", table[1:]))
+		for tableName := range tableNames {
+			if !tableRegex.MatchString(tableName) {
```
Comment (lance6716) [inline sync_diff_inspector/diff.go]: if match {delete}

File: sync_diff_inspector/diff.go (hunk 2 above)
Comment (lance6716) [reply sync_diff_inspector/diff.go]: I prefer `map[string]struct{}`

Comment (lichunzhu) [reply sync_diff_inspector/diff.go]: `map[string]struct{}` is easier to delete.

**Review by lance6716 — APPROVED** [review]

File: sync_diff_inspector/diff.go (hunk 1 above)
Comment (glorv) [reply sync_diff_inspector/diff.go:336]: LGTM，since the original structure is already map not slice

File: sync_diff_inspector/diff.go [hunk 4]
```diff
@@ -537,6 +535,24 @@ func (df *Diff) GetMatchTable(db DBConfig, schema, table string, allTables map[s
 	return tableNames, nil
 }
 
+// RemoveMatchTable removes all the matched table.
+func (df *Diff) RemoveMatchTable(table string, allTables map[string]struct{}) map[string]struct{} {
+	tableNames := allTables
+
+	if table[0] == '~' {
+		tableRegex := regexp.MustCompile(fmt.Sprintf("(?i)%s", table[1:]))
```
Comment (glorv) [inline sync_diff_inspector/diff.go]: How can we ensure the regexp patten is always a valid expression

Comment (glorv) [reply sync_diff_inspector/diff.go]: BTW, do we need to ensure the expression should match the whole table name by alway add "^$" to the expression？

Comment (lichunzhu) [reply sync_diff_inspector/diff.go]: > How can we ensure the regexp patten is always a valid expression

sync-diff will panic directly in the past. Now I have updated this part to report an error

Comment (lichunzhu) [reply sync_diff_inspector/diff.go]: > BTW, do we need to ensure the expression should match the whole table name by alway add "^$" to the expression？

I think we should. But dm doesn't have this rule either.
Maybe we could have a further discussion about this and do this in another PR? Current configuration doesn't violate the situation before: https://github.com/pingcap/tidb-tools/blob/915b22e4d42c1493edf90f1ccc79227e55e217ba/sync_diff_inspector/diff.go#L522

**Comment by lichunzhu** [comment]
Sync-diff 2.0 replace filter with `table_filter`, so this PR can be closed now.

**Pull request closed without merging**